In [28]:
import pandas as pd
import numpy as np
df = pd.read_csv("SampleSuperstore.csv")
df.head()

,Ship Mode,Segment,Country,City,State,Postal Code,Region,Category,Sub-Category,Sales,Quantity,Discount,Profit
0,Second Class,Consumer,United States,Henderson,Kentucky,42420,South,Furniture,Bookcases,261.9600,2,0.00,41.9136
1,Second Class,Consumer,United States,Henderson,Kentucky,42420,South,Furniture,Chairs,731.9400,3,0.00,219.5820
2,Second Class,Corporate,United States,Los Angeles,California,90036,West,Office Supplies,Labels,14.6200,2,0.00,6.8714
3,Standard Class,Consumer,United States,Fort Lauderdale,Florida,33311,South,Furniture,Tables,957.5775,5,0.45,-383.0310
4,Standard Class,Consumer,United States,Fort Lauderdale,Florida,33311,South,Office Supplies,Storage,22.3680,2,0.20,2.5164


In [29]:
df.shape

(9994, 13)

In [30]:
df.columns

Index(['Ship Mode', 'Segment', 'Country', 'City', 'State', 'Postal Code',
       'Region', 'Category', 'Sub-Category', 'Sales', 'Quantity', 'Discount',
       'Profit'],
      dtype='object')

In [31]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9994 entries, 0 to 9993
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Ship Mode     9994 non-null   object 
 1   Segment       9994 non-null   object 
 2   Country       9994 non-null   object 
 3   City          9994 non-null   object 
 4   State         9994 non-null   object 
 5   Postal Code   9994 non-null   int64  
 6   Region        9994 non-null   object 
 7   Category      9994 non-null   object 
 8   Sub-Category  9994 non-null   object 
 9   Sales         9994 non-null   float64
 10  Quantity      9994 non-null   int64  
 11  Discount      9994 non-null   float64
 12  Profit        9994 non-null   float64
dtypes: float64(3), int64(2), object(8)
memory usage: 1015.1+ KB


In [32]:
df.isnull().sum()


,0
Ship Mode,0
Segment,0
Country,0
City,0
State,0
Postal Code,0
Region,0
Category,0
Sub-Category,0
Sales,0


In [33]:
df.duplicated().sum()

np.int64(17)

In [34]:
df[df.duplicated(keep=False)].sort_values(by=df.columns.tolist())

,Ship Mode,Segment,Country,City,State,Postal Code,Region,Category,Sub-Category,Sales,Quantity,Discount,Profit
6409,First Class,Consumer,United States,Houston,Texas,77041,Central,Office Supplies,Paper,47.952,3,0.2,16.1838
8032,First Class,Consumer,United States,Houston,Texas,77041,Central,Office Supplies,Paper,47.952,3,0.2,16.1838
5493,Same Day,Home Office,United States,San Francisco,California,94122,West,Office Supplies,Labels,41.400,4,0.0,19.8720
5905,Same Day,Home Office,United States,San Francisco,California,94122,West,Office Supplies,Labels,41.400,4,0.0,19.8720
2807,Second Class,Consumer,United States,Seattle,Washington,98115,West,Office Supplies,Paper,12.960,2,0.0,6.2208
8095,Second Class,Consumer,United States,Seattle,Washington,98115,West,Office Supplies,Paper,12.960,2,0.0,6.2208
8457,Second Class,Corporate,United States,Chicago,Illinois,60653,Central,Office Supplies,Binders,3.564,3,0.8,-6.2370
9477,Second Class,Corporate,United States,Chicago,Illinois,60653,Central,Office Supplies,Binders,3.564,3,0.8,-6.2370
8533,Standard Class,Consumer,United States,Detroit,Michigan,48227,Central,Furniture,Chairs,389.970,3,0.0,35.0973
9262,Standard Class,Consumer,United States,Detroit,Michigan,48227,Central,Furniture,Chairs,389.970,3,0.0,35.0973


In [35]:
df.shape[0] - df.duplicated().sum()

np.int64(9977)

In [36]:
df.duplicated().sum()

np.int64(17)

In [37]:
df.shape

(9994, 13)

In [38]:
df = df.drop_duplicates()

In [39]:
df.duplicated().sum()

np.int64(0)

In [40]:
df.shape

(9977, 13)

In [41]:
df["Ship Mode"].unique()

array(['Second Class', 'Standard Class', 'First Class', 'Same Day'],
      dtype=object)

In [42]:
df["Segment"].value_counts()

,count
Segment,
Consumer,5183
Corporate,3015
Home Office,1779


In [43]:
df["Category"].value_counts()

,count
Category,
Office Supplies,6012
Furniture,2118
Technology,1847


In [44]:
df["Sub-Category"].value_counts()

,count
Sub-Category,
Binders,1522
Paper,1359
Furnishings,956
Phones,889
Storage,846
Art,795
Accessories,775
Chairs,615
Appliances,466


In [45]:
df["Sub-Category"].unique()

array(['Bookcases', 'Chairs', 'Labels', 'Tables', 'Storage',
       'Furnishings', 'Art', 'Phones', 'Binders', 'Appliances', 'Paper',
       'Accessories', 'Envelopes', 'Fasteners', 'Supplies', 'Machines',
       'Copiers'], dtype=object)

In [46]:
df["Sales"].describe()

,Sales
count,9977.000000
mean,230.148902
std,623.721409
min,0.444000
25%,17.300000
50%,54.816000
75%,209.970000
max,22638.480000


In [50]:
print(df.columns.tolist())

['Ship Mode', 'Segment', 'Country', 'City', 'State', 'Postal Code', 'Region', 'Category', 'Sub-Category', 'Sales', 'Quantity', 'Discount', 'Profit']


In [51]:
df["Sales"].describe()

,Sales
count,9977.000000
mean,230.148902
std,623.721409
min,0.444000
25%,17.300000
50%,54.816000
75%,209.970000
max,22638.480000


In [57]:
df["Quantity"].describe()

,Quantity
count,9977.000000
mean,3.790719
std,2.226657
min,1.000000
25%,2.000000
50%,3.000000
75%,5.000000
max,14.000000


In [58]:
df["Postal Code"].describe()

,Postal Code
count,9977.000000
mean,55154.964117
std,32058.266816
min,1040.000000
25%,23223.000000
50%,55901.000000
75%,90008.000000
max,99301.000000


In [59]:
print("Negative Sales:", (df["Sales"] < 0).sum())
print("Negative Quantity:", (df["Quantity"] < 0).sum())
print("Negative Discount:", (df["Discount"] < 0).sum())
print("Negative Profit:", (df["Profit"] < 0).sum())

Negative Sales: 0
Negative Quantity: 0
Negative Discount: 0
Negative Profit: 1869


In [60]:
df.duplicated().sum()

np.int64(0)

In [61]:
print("Ship Mode:")
print(df["Ship Mode"].value_counts())

print("\nSegment:")
print(df["Segment"].value_counts())

print("\nCountry:")
print(df["Country"].value_counts())

print("\nRegion:")
print(df["Region"].value_counts())

print("\nCategory:")
print(df["Category"].value_counts())

print("\nSub-Category:")
print(df["Sub-Category"].value_counts())

Ship Mode:
Ship Mode
Standard Class    5955
Second Class      1943
First Class       1537
Same Day           542
Name: count, dtype: int64

Segment:
Segment
Consumer       5183
Corporate      3015
Home Office    1779
Name: count, dtype: int64

Country:
Country
United States    9977
Name: count, dtype: int64

Region:
Region
West       3193
East       2845
Central    2319
South      1620
Name: count, dtype: int64

Category:
Category
Office Supplies    6012
Furniture          2118
Technology         1847
Name: count, dtype: int64

Sub-Category:
Sub-Category
Binders        1522
Paper          1359
Furnishings     956
Phones          889
Storage         846
Art             795
Accessories     775
Chairs          615
Appliances      466
Labels          363
Tables          319
Envelopes       254
Bookcases       228
Fasteners       217
Supplies        190
Machines        115
Copiers          68
Name: count, dtype: int64


In [63]:
df["Country"].value_counts(dropna=False)

,count
Country,
United States,9977


In [64]:
df["Country"].isna().sum()

np.int64(0)

In [65]:
df["Country"].eq("").sum()

np.int64(0)

In [66]:
df["Country"].astype(str).str.strip().eq("").sum()

np.int64(0)

In [67]:
df["Country"].unique()

array(['United States'], dtype=object)

In [68]:
df["Postal Code"].head(10)

,Postal Code
0,42420
1,42420
2,90036
3,33311
4,33311
5,90032
6,90032
7,90032
8,90032
9,90032


In [69]:
df["Postal Code"].nunique()

631

In [70]:
df["Postal Code"].isna().sum()

np.int64(0)

In [71]:
text_columns = df.select_dtypes(include="object").columns

for col in text_columns:
    spaces = (df[col] != df[col].str.strip()).sum()
    print(col, ":", spaces, "values with extra spaces")

Ship Mode : 0 values with extra spaces
Segment : 0 values with extra spaces
Country : 0 values with extra spaces
City : 0 values with extra spaces
State : 0 values with extra spaces
Region : 0 values with extra spaces
Category : 0 values with extra spaces
Sub-Category : 0 values with extra spaces


In [72]:
for col in text_columns:
    print("\n", col)
    print(df[col].unique()[:20])


 Ship Mode
['Second Class' 'Standard Class' 'First Class' 'Same Day']

 Segment
['Consumer' 'Corporate' 'Home Office']

 Country
['United States']

 City
['Henderson' 'Los Angeles' 'Fort Lauderdale' 'Concord' 'Seattle'
 'Fort Worth' 'Madison' 'West Jordan' 'San Francisco' 'Fremont'
 'Philadelphia' 'Orem' 'Houston' 'Richardson' 'Naperville' 'Melbourne'
 'Eagan' 'Westland' 'Dover' 'New Albany']

 State
['Kentucky' 'California' 'Florida' 'North Carolina' 'Washington' 'Texas'
 'Wisconsin' 'Utah' 'Nebraska' 'Pennsylvania' 'Illinois' 'Minnesota'
 'Michigan' 'Delaware' 'Indiana' 'New York' 'Arizona' 'Virginia'
 'Tennessee' 'Alabama']

 Region
['South' 'West' 'Central' 'East']

 Category
['Furniture' 'Office Supplies' 'Technology']

 Sub-Category
['Bookcases' 'Chairs' 'Labels' 'Tables' 'Storage' 'Furnishings' 'Art'
 'Phones' 'Binders' 'Appliances' 'Paper' 'Accessories' 'Envelopes'
 'Fasteners' 'Supplies' 'Machines' 'Copiers']


In [73]:
text_columns = df.select_dtypes(include="object").columns

for col in text_columns:
    spaces = (df[col] != df[col].str.strip()).sum()
    print(col, ":", spaces, "values with extra spaces")

Ship Mode : 0 values with extra spaces
Segment : 0 values with extra spaces
Country : 0 values with extra spaces
City : 0 values with extra spaces
State : 0 values with extra spaces
Region : 0 values with extra spaces
Category : 0 values with extra spaces
Sub-Category : 0 values with extra spaces


In [75]:
cleaned_df = df.copy()

In [77]:
cleaned_df = cleaned_df.drop_duplicates()

In [78]:
text_columns = cleaned_df.select_dtypes(include="object").columns

for col in text_columns:
    cleaned_df[col] = cleaned_df[col].str.strip()

In [79]:
numeric_columns = ["Postal Code", "Sales", "Quantity", "Discount", "Profit"]

for col in numeric_columns:
    cleaned_df[col] = pd.to_numeric(cleaned_df[col], errors="coerce")

In [80]:
cleaned_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 9977 entries, 0 to 9993
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Ship Mode     9977 non-null   object 
 1   Segment       9977 non-null   object 
 2   Country       9977 non-null   object 
 3   City          9977 non-null   object 
 4   State         9977 non-null   object 
 5   Postal Code   9977 non-null   int64  
 6   Region        9977 non-null   object 
 7   Category      9977 non-null   object 
 8   Sub-Category  9977 non-null   object 
 9   Sales         9977 non-null   float64
 10  Quantity      9977 non-null   int64  
 11  Discount      9977 non-null   float64
 12  Profit        9977 non-null   float64
dtypes: float64(3), int64(2), object(8)
memory usage: 1.1+ MB


In [81]:
print("Rows:", cleaned_df.shape[0])
print("Columns:", cleaned_df.shape[1])
print("Duplicate rows:", cleaned_df.duplicated().sum())
print("Total missing values:", cleaned_df.isnull().sum().sum())

Rows: 9977
Columns: 13
Duplicate rows: 0
Total missing values: 0


In [82]:
cleaned_df.to_csv("superstore_cleaned.csv", index=False)

In [83]:
import os

print(os.path.exists("superstore_cleaned.csv"))

True


In [84]:
final_df = pd.read_csv("superstore_cleaned.csv")

print("========== FINAL DATASET VALIDATION ==========")
print("Rows:", final_df.shape[0])
print("Columns:", final_df.shape[1])
print("Duplicate rows:", final_df.duplicated().sum())
print("Total missing values:", final_df.isnull().sum().sum())

========== FINAL DATASET VALIDATION ==========
Rows: 9977
Columns: 13
Duplicate rows: 0
Total missing values: 0


In [85]:
final_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9977 entries, 0 to 9976
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Ship Mode     9977 non-null   object 
 1   Segment       9977 non-null   object 
 2   Country       9977 non-null   object 
 3   City          9977 non-null   object 
 4   State         9977 non-null   object 
 5   Postal Code   9977 non-null   int64  
 6   Region        9977 non-null   object 
 7   Category      9977 non-null   object 
 8   Sub-Category  9977 non-null   object 
 9   Sales         9977 non-null   float64
 10  Quantity      9977 non-null   int64  
 11  Discount      9977 non-null   float64
 12  Profit        9977 non-null   float64
dtypes: float64(3), int64(2), object(8)
memory usage: 1013.4+ KB


# Data Cleaning and Preprocessing – Sample Superstore

## Project Objective

The objective of this project is to clean and preprocess the Sample Superstore dataset using Python and Pandas. The dataset was inspected for missing values, duplicate records, inconsistent text formatting, incorrect data types, and invalid numerical values.

## Tools Used

- Python
- Pandas
- NumPy
- Google Colab

## Dataset Overview

The original dataset contains 9,994 rows and 13 columns. It contains information about shipping mode, customer segment, location, product categories, sales, quantity, discount, and profit.

## Data Quality Checks

The following checks were performed:

1. Dataset structure and dimensions
2. Missing values
3. Duplicate records
4. Categorical/text consistency
5. Extra whitespace
6. Numerical data types
7. Negative numerical values
8. Final validation

## Cleaning Performed

- Duplicate rows were removed.
- Text columns were checked and stripped of leading/trailing whitespace.
- Numeric columns were converted/validated as numeric data types.
- Missing values were checked.
- Categorical fields were checked for inconsistent formatting.
- Negative profit values were retained because they represent legitimate loss-making transactions.

## Final Result

The cleaned dataset contains 9,977 rows and 13 columns.

Final validation confirmed:

- 0 duplicate rows
- 0 missing values
- 13 columns
- Correct numeric data types

The cleaned dataset was exported as `superstore_cleaned.csv`.

# Data Cleaning Change Log

| Check / Issue | Finding | Action Taken | Final Result |
|---|---|---|---|
| Missing values | No missing values found | No imputation required | 0 missing values |
| Duplicate records | 17 duplicate rows identified during cleaning | Removed duplicate rows | 0 duplicates |
| Text formatting | Text fields validated and stripped | Consistent formatting |
| Category | Consistent values | No correction required | Valid |
| Segment | Consistent values | No correction required | Valid |
| Sub-Category | Consistent values | No correction required | Valid |
| Sales | No negative values | Validated | Valid |
| Quantity | No negative values | Validated | Valid |
| Discount | No negative values | Validated | Valid |
| Profit | 1,869 negative values | Retained because they represent legitimate losses | Valid |
| Numeric data types | Numeric columns correctly typed | Validated/converted | Correct |
| Final dataset | 9,994 original rows | Duplicate removal | 9,977 rows × 13 columns |

## Conclusion

The Sample Superstore dataset was successfully cleaned and validated using Python and Pandas. The final dataset contains 9,977 rows and 13 columns, with no missing values or duplicate rows. Negative profit values were retained because they represent legitimate loss-making transactions.

In [54]:
df["Discount"].describe()

,Discount
count,9977.000000
mean,0.156278
std,0.206455
min,0.000000
25%,0.000000
50%,0.200000
75%,0.200000
max,0.800000


In [56]:
df["Profit"].describe()

,Profit
count,9977.00000
mean,28.69013
std,234.45784
min,-6599.97800
25%,1.72620
50%,8.67100
75%,29.37200
max,8399.97600
